lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -358,6 +361,13 @@ LookupTypeResult TypeChecker::lookupMemberType(DeclContext *dc,
       }
     }
 
+    // Ignore typealiases found in protocol members.
+    if (auto alias = dyn_cast<TypeAliasDecl>(typeDecl)) {
+      auto aliasParent = alias->getParent();
+      if (aliasParent != dc && isa<ProtocolDecl>(aliasParent))
+        continue;
+    }
```

Comment:
This part shouldn't go in as-is. The motivation is for situations like:
```
protocol P1 {
  associatedtype A
}
protocol P2 {
  associatedtype P : P1
  typealias A = P.A
}
struct S: P1, P2 {
}
```
We create an implicit `typealias A = whatever` in `struct S` to record the chosen witness type, but then since protocol members are visible in lookup, type checking gets very confused because it has an overload between the `whatever` and `P.A` declarations everywhere `A` gets checked.

Ideally we'd avoid choosing overloads in this situation, always choosing the bound associatedtype, but still leaving `A = P.A` visible to nominals that conform to P2 without P1 (or any other protocol that has an `associatedtype A`).